### ISSUE TITLE
More frequent, consistent, and predictable feature updates

### ISSUE THREAD

**Issue by ninavizz** [QubesOS/qubes-issues#6515]
In a recent ticket (that I cannot find) it was mentioned that there's a rule for Qubes releases: point releases being backwards-compatible security updates, and full releases being the only releases to include GUI feature updates. 

In our recent user survey, many users requested more frequent feature releases. Likewise, more visibility into what is in-progress and to be expected at what point. 

I am also observing our soon-to-deploy `4.1` release burdened by a desire to squeeze-in "allthethings!," rather than focus on pushing-out a few significant, service-based or structural, and not-trivial to debug features, with an updated dom0 version of Fedora, all bug-free and tested thoroughly.  

In order to ship GUI features more quickly, I wanted to propose that we change our release criteria to allow for user-facing GUI components to be included in point-releases, with @marmarek setting the criteria for each major release around deeper architectural updates that are likely to also require more testing. The [four bullet-points in the recent 4.1 release article](https://www.qubes-os.org/news/2020/06/22/new-qrexec-policy-system/), felt like almost too-significant of a release—to be honest. For users. Those items alone, are a meaty release for users to get their heads around. GUI stuff should deploy on its own—as much as our complex testing needs can allow for.

Unfortunately I cannot find the release numbering/criteria stuff at all, anyhere in the docs—so apologize for not speaking to that text, directly.

**Comment by DemiMarie** [comment]
I second this!

**Comment by andrewdavidwong** [comment]
> In order to ship GUI features more quickly, I wanted to propose that we change our release criteria to allow for user-facing GUI components to be included in point-releases

No. It's not like we randomly decided that point releases should contain only backward-compatible bug fixes. This is the standard way software versioning is done. See https://semver.org/ (which I also linked in the [other issue](https://github.com/QubesOS/qubes-issues/issues/6487#issuecomment-809448269)). If you want more frequent feature updates, why not just have more frequent minor and major releases? Trying to achieve that by changing the meanings of version numbers is like trying to get more food out of a pizza by cutting it up into more slices.

The real ask here is for more frequent feature releases, not for changing version scheme in any way.

**Comment by ninavizz** [comment]
> Trying to get more feature releases by changing the meanings of version numbers is like trying to get more food out of a pizza by cutting it up into more slices.

What, you don't do that? ;)

On topic, though... so, yes, on one hand, you're correct. A big distinction between Qubes and other products I've worked on, however, is that when dom0's version of Fedora and major architectural things go into releases, they can take FOR EVER to debug and to get right. Trying to hasten those releases, would be a recipe for disaster. 

So, another idea: it seems like the above-type release, can take about 12-18mos to achieve. Which is about the cadence Apple has, for entirely-new versions of MacOS... if not longer. 

@marmarek @marmarta @andrewdavidwong @mfc What do y'all think about potentially working on two releases at once, simultaneously; planning that odd-numbered releases to be major-architectural features and new dom0 versions of Fedora, and maaaybe some GUI stuff, and then even-numbered releases GUI stuff only. That way we could probably do more predictable release cycles, smaller releases, and one "class" of release would be gnarly with testing and debugging, while the other far more lightweight?

In reflecting upon all past software orgs I've worked with, I'm also now realizing that it was a common thing to work on 2-3 releases at once; each staggered like this, to get more predictable and more frequent releases out the door.

**Comment by andrewdavidwong** [comment]
> On topic, though... so, yes, you are correct. A big distinction between Qubes and other products I've worked on, however, is that when dom0's version of Fedora and major architectural things go into releases, they can take FOR EVER to debug and to get right across the zillions of hardware allthethings. Trying to hasten those releases, would be a recipe for disaster (until we get that multi-million dollar QA lab funded and we could QA and fix everything in a few days, of course, hehe).
> 

Why would those major releases have to be hastened? Why not just do as many minor (not patch/point) releases as we want in between the major releases and not rush the major releases?


> So, another idea: it seems like the above-type release, can take about 12-18mos to achieve. Which is about the cadence Apple has, for entirely-new versions of MacOS... if not longer. 
> 
> 
> 
> @marmarek @marmarta @andrewdavidwong @mfc What do y'all think about potentially working on two releases at once, simultaneously; planning that odd-numbered releases to be major-architectural features and new dom0 versions of Fedora, and maaaybe some GUI stuff, and then even-numbered releases GUI stuff only. That way we could probably do more predictable release cycles, smaller releases, and one "class" of release would be gnarly with testing and debugging, while the other far more lightweight?
> 

Again, why not just do this with minor releases? Why try to turn half of major releases into minor releases when we already have minor releases? Unless I'm missing something, it sounds like the GUI stuff would be appropriately backward-compatible, hence suitable for minor releases.

> In reflecting upon all past software orgs I've worked with, I'm also now realizing that it was a common thing to work on 2-3 releases at once; each staggered like this, to get more predictable and more frequent releases out the door.

We sort of do that already, but the main reason it doesn't happen more is probably limited developer bandwidth.

**Comment by ninavizz** [comment]
QubesOS does not currently adhere to a schedule of releases. They just happen/ship, when they're ready. That's not working for the broader community, and it won't work for more mainstream users (vs just FOSS enthusiasts). That needs to change. I don't care how it changes. I've been observing `4.1` feature-creep and waffle for months, wanting to pull my hair out all the while in frustration. While also observing Marek work all hours, and everyone else busting their tails. There's no laziness. Just poor meeting of user needs via expectations management. Which I say with love and admiration and reverence for all of the hard work that is and has been happening, and I trust will continue to happen. There is no lack of hard work happening. 

Many at-risk users need this product. It will never hit the ubiquity of Windows or even Ubuntu, but its current abandonment rate and adoption rates feel problematic. Does the core team desire QubesOS to remain a niche OS limited to use by the FOSS enthusiast community, or does the core team want QubesOS to be more accessible to users who might sincerely depend on it (outside of developers and security researchers)? QubesOS also needs funding, to achieve the usability I'd personally like to see happen... and funders want predictability and to see things happen more quickly. Not hard things, just small and more frequent things.

I created this Issue to take a positive step, towards enabling such change to happen. Especially because there is such limited developer bandwidth, something or a lot of somethings, need to be done differently.

**Comment by DemiMarie** [comment]
I wholeheartedly agree with @ninavizz.  Having small changes held up on major features is an anti-pattern, IMO.

**Comment by ninavizz** [comment]
When users have zero visibility into planned improvements, they'll bail instead of "toughing it out." Likewise, when they just never see noticeable or meaningful improvements to their daily experience with a product... yet are constantly running security releases. The "why" that happens, is completely understandable. Yet that doesn't change the poor experience for users. 

A disciplined approach to keeping releases smaller and more frequent; to making public date-estimates for delivery and then reducing scope as-needed while being more conservative when scoping things at the outset; and working on multiple releases at once to ensure releases are kept smaller and deployed with user-facing updates more often, could be some good starting points. Which I feel like a jerk, saying, because all that implies that folks simply doesn't care or aren't currently trying—which I know for a fact, to not be the case. At. All. <3

**Comment by ninavizz** [comment]
@andrewdavidwong In [the other issue](https://github.com/QubesOS/qubes-issues/issues/6487#issuecomment-809448269) (TY for linking it to me, btw!) you said that point-releases (0.1.xyz) are for backwards-compatible bug fixes only. Not new features. 

My ask in this ticket, is to differentiate between QA-intensive features, and small features; and to bundle them into different releases, then planning for smaller feature releases to be deployed on a quarterly basis, with the biggy architectural and gnarly to QA releases, happening every 12-24 months. 

Users have made it very clear in the +4,500 person survey, that they want to see usability improvements to Qubes' GUI. Even with our radically-small/lean team, we should be able to push-out new GUI features, iterative improvements to existing widgets, and all-new widgets, more often, and more predictably. How to enable that, is my only goal with this.

**Comment by h01ger** [comment]
On Fri, Apr 09, 2021 at 11:13:24AM -0700, Nina Eleanor Alter wrote:
> QubesOS does not currently adhere to a schedule of releases. They just
> happen/ship, when they're ready. That's not working for the broader
> community, and it won't work for more mainstream users (vs just FOSS
> enthusiasts). That needs to change. I don't care how it changes.

I agree (mostly) with your observation but less with your conclusions
(or the conclusion from it, "I don't care how"..) and while I agree
I also fear what impact this harsh framing has on those doing the
actual development work. Because I also believe they will agree and
are not to blame, so instead of telling them what to do I believe the
circumstances need to be improved.

"More frequent feature releases" thus to me is an result of the changes
needed, but in itself not the change needed nor the problem to be addressed
primarily. it's a symptom.


-- 
cheers,
	Holger

 ⢀⣴⠾⠻⢶⣦⠀
 ⣾⠁⢠⠒⠀⣿⡁       holger@(debian|reproducible-builds|layer-acht).org
 ⢿⡄⠘⠷⠚⠋⠀ PGP fingerprint: B8BF 5413 7B09 D35C F026 FE9D 091A B856 069A AA1C
 ⠈⠳⣄

**Comment by ninavizz** [comment]
> I agree (mostly) with your observation but less with your conclusions
> (or the conclusion from it, "I don't care how"..) and while I agree
> I also fear what impact this harsh framing has on those doing the
> actual development work. 

TY for that @h0lger. "I don't care how" was not a good choice of words, I agree. A more accurate choice of words, would be simply stating that the mechanics of "how" are less of a concern to me, than meeting the goal itself. My intent is not to dictate a solution—yet I also don't want to scream "This is a problem, fix it!." 

> "More frequent feature releases" thus to me is an result of the changes
> needed, but in itself not the change needed nor the problem to be addressed
> primarily. it's a symptom.

Also wholeheartedly agreed. The problems in need of addressing needs to be clearly identified, and changes needed shaped from that. That would have been better phrasing for me to have used in my initial response to @andrewdavidwong. Whom I have all the respect for. As well as every developer pushing lines of code to make Qubes happen, and the non-developers busting tail to keep the project on track. 

I'd also love to do a structured workshop to try and get to the root of some problem/solution things, but appreciate that many would be apprehensive to a synchronous meeting. If folks would be open to or interested in a synchronous or asynchronous workshop, perhaps that could be a starting place?

**Comment by andrewdavidwong** [comment]
> @andrewdavidwong In [the other issue](https://github.com/QubesOS/qubes-issues/issues/6487#issuecomment-809448269) (TY for linking it to me, btw!) you said that point-releases (0.1.xyz) are for backwards-compatible bug fixes only. Not new features. "Iterative improvements to existing features" are neither bugs, nor new features—so where do they fit?

Improvements to existing features would generally go in minor releases ("4.1," "4.2," etc.). However, this version numbering thing is probably the least important aspect of this general topic. (It's the slicing of the pizza rather than the growing, cooking, or eating.)

-----

If I understand correctly, it looks like we've arrived at the real ask being more frequent, consistent, and predictable feature updates (meaning improvements to existing features as well as new features, and not just security updates). In particular, the desire is to avoid feature creep and avoid having unrelated updates block each other, which you say you've been seeing happen a lot with 4.1.

In principle, I agree that this would be good, but I also know that the perspective from the outside of the development process can be quite different from the perspective from the inside. Are there special reasons that this has been happening with 4.1? Would it be costly to avoid for some non-obvious reason? I'd be curious to hear from @marmarek on these points, as I believe only he can authoritatively answer them.

**Comment by ninavizz** [comment]
> I also know that the perspective from the outside of the development process can be quite different from the perspective from the inside. Are there special reasons that this has been happening with 4.1? Would it be costly to avoid for some non-obvious reason? I'd be curious to hear from @marmarek on these points, as I believe only he can authoritatively answer them.

1000% (which I can say because I have an art degree)

This has been a hard topic to bring-up, as _nobody_ is "at fault" for anything. And it is very, very hard to frame and evaluate, without implying that.

**Comment by h01ger** [comment]
On Fri, Apr 09, 2021 at 02:41:52PM -0700, Nina Eleanor Alter wrote:
> TY for that @h0lger. "I don't care how" was not a good choice of words, I agree. 
> [...] 
> Also wholeheartedly agreed.
> [...]

Thanks for acknowledging, reflecting and moving on, improving on it. (the stuff
not quoted here.) wow, nice, very.


-- 
cheers,
	Holger

 ⢀⣴⠾⠻⢶⣦⠀
 ⣾⠁⢠⠒⠀⣿⡁       holger@(debian|reproducible-builds|layer-acht).org
 ⢿⡄⠘⠷⠚⠋⠀ PGP fingerprint: B8BF 5413 7B09 D35C F026 FE9D 091A B856 069A AA1C
 ⠈⠳⣄

**Comment by GWeck** [comment]
Perhaps one more aspect to consider: Looking at the way from R3.2 to R4.0 and now to R4.1, there has been a lot of (excellent) internal resstructuring, together with improving UX and minor funtionalities. In some parts, these internal changes have broken existing, mainly community supported functionality like Qubes Windows Tools, which may be essential for users coming from Windows and desperately needing a secure environment. For these users, it took some time until the needed functionality was there again or will be (for R4.1) restored. This forced them to stay with the older version until the missing parts worked again.

Perhaps this problem could be mitigated by releasing non-structural improvements, like UX enhancements or supporting new drivers, more often, maybe in point releases like R4.0.x, while concentrating structural changes in major releases like R4.x. So the overall functionality could be cautiously enhanced while keeping the internal structure, and thus continuing the support of things like QWT. For users, going to a new major release could thus be postponed until the needed environment is ready, but they still would get an improved usage experience. (Being forced to use Windows for some work, I have no problem staying with older versions of a system, as long as they accomplish what I need. So staying with a functionally improved version R4.0 for several years would still suit me fine.)

**Comment by marmarek** [comment]
Generally I agree we need more frequent feature releases. But I don't like to differentiate between "UX improvements releases" and "major features" on the process level. This means we'd indeed need to develop several (two at least) feature releases in parallel, which is very hard to maintain with our limited team. Generally I think the main issue here is R4.1 being overdue for about a year already, which has two major reasons:
 - merging some features too early (before stabilizing them enough), resulting in #6174 for example
 - 2020 being hard and full of surprises (some positive, but still taking significant amount of my time)

The first one we try to solve with extending our testing workflow, especially automated tests (if you look closely, #6174 was identified by our improved tests, just a bit too late).

For the second one, it's really about the team size, effectively too low bus factor. In practice, among other roles, I'm release manager for both upcoming R4.1, and for stable R4.0. And also I have some responsibilities in other projects (public and non public). I think I could use some help in any of those areas, but since they are connected with rather high responsibility, it isn't exactly easy (for me) to delegate here...

**Comment by andrewdavidwong** [comment]
> Perhaps this problem could be mitigated by releasing non-structural improvements, like UX enhancements or supporting new drivers, more often, maybe in point releases like R4.0.x, while concentrating structural changes in major releases like R4.x. So the overall functionality could be cautiously enhanced while keeping the internal structure, and thus continuing the support of things like QWT.

Please read above. We have already covered this repeatedly. Point/patch releases are for backward-compatible bug fixes. Minor releases are for backward-compatible enhancements. Major releases are for backward-incompatible changes. Half of what you're saying is the whole point of this issue, and the other half we've already been doing.

> For users, going to a new major release could thus be postponed until the needed environment is ready, but they still would get an improved usage experience. (Being forced to use Windows for some work, I have no problem staying with older versions of a system, as long as they accomplish what I need. So staying with a functionally improved version R4.0 for several years would still suit me fine.)

The problem is that we can't just keep supporting old releases until everyone is ready (which would be indefinitely), since then we would have no time to work on new stuff. That's why we established a support schedule: https://www.qubes-os.org/doc/supported-versions/#qubes-os

-----

> Generally I agree we need more frequent feature releases. But I don't like to differentiate between "UX improvements releases" and "major features" on the process level. This means we'd indeed need to develop several (two at least) feature releases in parallel, which is very hard to maintain with our limited team.

I agree. I don't see the need for it, when we could instead just have more frequent minor releases, where each minor release contains whatever's ready at the time from among the "UX improvement" and "major features" buckets.

> Generally I agree we need more frequent feature releases. But I don't like to differentiate between "UX improvements releases" and "major features" on the process level. This means we'd indeed need to develop several (two at least) feature releases in parallel, which is very hard to maintain with our limited team. Generally I think the main issue here is R4.1 being overdue for about a year already, which has two major reasons:
> 
> * merging some features too early (before stabilizing them enough), resulting in #6174 for example
> * 2020 being hard and full of surprises (some positive, but still taking significant amount of my time)
> 
> The first one we try to solve with extending our testing workflow, especially automated tests (if you look closely, #6174 was identified by our improved tests, just a bit too late).
> 

I think another problem @ninavizz has identified is a tendency among many people to want to pack as much possible into 4.1, perhaps because there's this underlying fear that minor releases are so infrequent that if a feature misses the 4.1 boat, we'll have to wait an awfully long time until 4.2. Once again, this could be fixed by having more frequent minor releases. If we establish that there will be a new minor release every N months, regardless of what's ready to go into it, then some will be large and some will be small, and some will contain significant new features while others contain mere iterative improvements, but at least there will be no starvation instinct that pushes people to gorge themselves on scarce releases, since we all know the next one is right around the corner.

> For the second one, it's really about the team size, effectively too low bus factor. In practice, among other roles, I'm release manager for both upcoming R4.1, and for stable R4.0. And also I have some responsibilities in other projects (public and non public). I think I could use some help in any of those areas, but since they are connected with rather high responsibility, it isn't exactly easy (for me) to delegate here...

I concur with the diagnosis, but I can offer no cure. :slightly_frowning_face:

**Comment by ninavizz** [comment]
> >For the second one, it's really about the team size, effectively too low bus factor. In practice, among other roles, I'm release manager for both upcoming R4.1, and for stable R4.0. And also I have some responsibilities in other projects (public and non public). I think I could use some help in any of those areas, but since they are connected with rather high responsibility, it isn't exactly easy (for me) to delegate here...
> 
> I concur with the diagnosis, but I can offer no cure. 🙁

I probably should have waited until 4.1 got deployed, to offer this issue up to just the core team. I'm so sorry. Still learning the best ways to communicate with a fully async team. You're all amazing and wonderful; Marek, you already work far too hard, and 2020 has indeed been quite a gut-punch. 2021 isn't doing a much better job improving upon its predacessor, tbh... so I think it's ok if we all similarly postpone this thread, until at least the world around us decides to be better. 

> I think another problem @ninavizz has identified is a tendency among many people to want to pack as much possible into 4.1, perhaps because there's this underlying fear that minor releases are so infrequent that if a feature misses the 4.1 boat, we'll have to wait an awfully long time until 4.2. Once again, this could be fixed by having more frequent minor releases. If we establish that there will be a new minor release every N months, regardless of what's ready to go into it, then some will be large and some will be small, and some will contain significant new features while others contain mere iterative improvements, but at least there will be no starvation instinct that pushes people to gorge themselves on scarce releases, since we all know the next one is right around the corner.

This, I actually 100% agree with and think could be a terrific, simple first step. Baby steps are ok, nobody 'round here needs to try and be Atlas for this one. Just trying to ensure we're not losing what we've learned in our first major research effort.

**Comment by GWeck** [comment]
@andrewdavidwong Maybe you got me wrong: I am aware of the current updating scheme, but I want to point out that, from the perspective of a general, non-technical user, there is a problem with this scheme. I you look at the current R4.0.4, this system just looks like the original R4.0. So, nothing seems to have happened in the last two years. On the other hand, the R4.1, which contains just a lot of wonderful stuff, both structural and functional, is still in development. In order to improve this external  impression, an intermediate level of update might help. (Surely not for the way to R4.1, but afterwards! According to the current tests, the R4.1 is very well on its way, and I am deeply impressed by what such a rather small team has achieved and is achieving - a quality far better than what you get for a lot of money from the big software firms.)

For the time after R4.1, it might improve the way the system is looked upon if smaller functional updates based on the same internal structure could be released more often, while the larger structural enhancements could be kept together until some new major version. Let me explain this at the example of the Qube manager. For R4.1, this tool is significantly faster than the R4.0 version, probably because of the internal workings of getting the information from `qvm-ls`. Maybe this required using some improved internal interfaces which are only available in R4.1. On the other hand, the design of this tool has been improved by better icons, new filtering and search possibilities, and so on. Showing this improved design to the external user would just create the impression that the system implementation is permanently moving forward.

I am well aware that the available resources are limited and that having two development versions - a functional one and a structural one - is probably out of scope. On the other hand, showing to possible future users that the system is improved all the time might help to get more people on the wagon. So it might help to put functional changes which can be implemeted without too much effort more often into an officially supported version - just in order to enhance the visibility of Qubes.

I would be glad if the system was used in more places, where currently people are struggling with things like Emotet, which could be dealt with much more efficiently in a Qubes environment! So, keep up the good work!

**Comment by andrewdavidwong** [comment]
@GWeck, I'm puzzled. Isn't that what I've been saying here all along? Please read my comments above again. I don't know how to be any clearer than I already have been. I feel like people are replying to me without actually reading what I'm writing and linking.

**Comment by GWeck** [comment]
@andrewdavidwong You`re absolutely right, but as you say, the main problem is resources, and that has to be solved on a different level.  I wish I knew a way to get more people interested in Qubes, beside telling anyone we can reach about it.

On the other hand, just increasing the team size by putting more people into development would probably increase the problem if the newcomers had not the level of expertise of the current team - see Microsoft - and getting good developers is a difficult task!

**Comment by 0spinboson** [comment]
nina: I personally find the lack of information on what's being worked on more of an issue than the lack of new features being released (though this is likely partly due to my running R4.1 basically since its public alpha/beta a year and a half or so ago).
I would agree that the team is probably adding too many features at once, but (speaking as a layman) given how much movement there's been in the past year on the hw support and Xen features front, I'm not really sure it was avoidable in this case either? Which again makes me circle back to the point that I think one way to create more of a community would be to tell folks more about the _process_ of developing Qubes OS. For a good example of this, see the way Keyboardio explained the issues they ran into during the model 01 production, on kickstarter. 
That said, the problem with that is that you need someone who a. can explain the technological challenges, and b. enjoys telling others about them, in a blog or vlog or whatever format. Which isn't easy, and so may or may not be unfeasible for Qubes. :)

**Comment by mfc** [comment]
> I think another problem @ninavizz has identified is a tendency among many people to want to pack as much possible into 4.1, perhaps because there's this underlying fear that minor releases are so infrequent that if a feature misses the 4.1 boat, we'll have to wait an awfully long time until 4.2. Once again, this could be fixed by having more frequent minor releases.

i think this describes a key problematic tendency well (i am certainly guilty of it), and it negatively impacts both the dev team and users. given we are now on minor releases, having a faster release schedule with less content would help alleviate this. 

having a clear & expected release schedule would help make everyone feel less guilty about excluding important features that are not quite done (but seem close).

the current [release schedule](https://www.qubes-os.org/doc/version-scheme/#release-schedule) is heavily focused on release candidates but from this thread and our general experience, users are more interested in understanding the general cadence of Qubes releases. maybe the Supreme Committee can convene to discuss :P

> I'm release manager for both upcoming R4.1, and for stable R4.0.

what about cycling between appropriate team members as release managers? or @marmarek you try to become release manager only for major releases? 

> also I have some responsibilities in other projects (public and non public). I think I could use some help in any of those areas, but since they are connected with rather high responsibility, it isn't exactly easy (for me) to delegate here...

this continues to be a problem with the project, but every time you are able to make explicit your responsibilities and successfully delegate them that is helpful and positively ripples throughout the project.

> problem with that model of course is that you need someone who a. can explain the technological challenges, and b. enjoys telling others about them

i could imagine someone from the community may be interested and have the requisite knowledge... @andrewdavidwong i don't think you have the time for this yourself and maybe it's worth us publicly asking the community. "communications coordinator"?

**Comment by andrewdavidwong** [comment]
> i could imagine someone from the community may be interested and have the requisite knowledge... @andrewdavidwong i don't think you have the time for this yourself and maybe it's worth us publicly asking the community. "communications coordinator"?

We occasionally do make "dev journal" type posts like this, but the process necessarily takes a lot of developers' time. They have to write out the technical content (I lack the knowledge to provide it myself), then I have to go through N rounds of revision with them to make it consumable by our readers. I consider this part of my job, so I'm more worried about their time than mine in doing this. The risk with getting the technical content from someone other than a core team member is that it may not be accurate. (IMO, it is very difficult to get this stuff right without specialized expertise *and* inside-track experience, so even many core dev posts are still checked by Marek.) This could be even *more* time-consuming for our core devs, since they'd have to correct every inaccuracy to the point where it might have been easier to start fresh with them as the technical source to begin with.

**Comment by ninavizz** [comment]
> @ninavizz : The lack of information on what's being worked on more personally bugs me more than the lack of new features being released (though this is likely partly due to my running R4.1 basically since its public alpha/beta a year and a half or so ago).
> 
> I on the one hand agree that the team is probably trying to add too many features at once, yet at the same time (and speaking as a layman), given how much movement there's been in the past year on the hardware support and Xen features front, and Qubes- reliance on Xen development, I'm not sure it was avoidable in this case either? 

One thing I have noticed as "lacking" that I had thought to bring up at some point among the core team, is developing a concrete product strategy (something I've done before) to identify priorities for the project. Such an artifact's goal would be to offer more flexibility than a set-in-stone roadmap, while also giving developers and project managers something to look at to prioritize things (to be worked on or pushed into a release ahead of others) against with more intent than what's currently being done. I mostly observed the need, because with user research findings need to "live" somewhere in an actionable fashion that folks can readily access and put to work with what features are prioritized over others. A project's Product Strategy, is typically that place.

With FOSS, obviously this gets complicated by upstream dependencies such as XEN and other things... but I think should be manageable. If folks would be interested in having such an artifact created to focus dev efforts and help prioritize release candidates more efficiently, I think this very thing would be a perfect effort to try an asynchronous workshop with. Which I'd design and moderate, and produce an end-artifact from. Thoughts?

**Comment by andrewdavidwong** [comment]
> > @ninavizz : The lack of information on what's being worked on more personally bugs me more than the lack of new features being released (though this is likely partly due to my running R4.1 basically since its public alpha/beta a year and a half or so ago).
> > I on the one hand agree that the team is probably trying to add too many features at once, yet at the same time (and speaking as a layman), given how much movement there's been in the past year on the hardware support and Xen features front, and Qubes- reliance on Xen development, I'm not sure it was avoidable in this case either?
> 
> One thing I have noticed as "lacking" that I had thought to bring up at some point among the core team, is developing a concrete product strategy (something I've done before) to identify priorities for the project. Such an artifact's goal would be to offer more flexibility than a set-in-stone roadmap, while also giving developers and project managers something to look at to prioritize things (to be worked on or pushed into a release ahead of others) against with more intent than what's currently being done. I mostly observed the need, because with user research findings need to "live" somewhere in an actionable fashion that folks can readily access and put to work with what features are prioritized over others. A project's Product Strategy, is typically that place.
> 
> With FOSS, obviously this gets complicated by upstream dependencies such as XEN and other things... but I think should be manageable. If folks would be interested in having such an artifact created to focus dev efforts and help prioritize release candidates more efficiently, I think this very thing would be a perfect effort to try an asynchronous workshop with. Which I'd design and moderate, and produce an end-artifact from. Thoughts?

I think it could be valuable, but it mainly depends on whether Marek and other core devs think they would actually use it to guide their decision-making, and that's a personal question that only they can answer. (I don't want to ping Marek right now, because I know he's trying to focus on pushing out 4.1. Perhaps you could ask him afterward.)

**Comment by ninavizz** [comment]
There are many great ideas in this thread to un-block general "What's happening & ahead?" visibility, but I am 100% with @andrewdavidwong in identifying "What does Marek think?" and getting a thoughtful Marek brain engaged on the topic, as a keystone to any possible action on it—while also not wanting to bother his focus from `4.1` just yet. 

AS SUCH I am going to close this issue right now, but then re-open it once `4.1` deploys—at which time we then make this a priority to address, only superseded by blocker or security bugs discovered post-release, after we've passed that critical marker. 

Lordie knows I don't want anyone burdened or feeling any more inadequate, than folks may already be feeling. 

## To Re-Open after 4.1 Deploys

**Comment by mfc** [comment]
> We occasionally do make "dev journal" type posts like this, but the process necessarily takes a lot of developers' time. [...]
> I consider this part of my job, so I'm more worried about their time than mine in doing this. 

I don't mean "dev journal" type posts, but rather very basic and informative post about the current state of the next release (where we are in the process) in normal user language (not in dev language). it is less even about the content than about maintaining a continuous connection via communication with the userbase. maybe starting with a post every quarter or something.

i think this connects with having a public project roadmap (https://github.com/QubesOS/qubes-issues/issues/2032).

> I am going to close this issue right now, but then re-open it once `4.1` deploys—at which time we then make this a priority to address, only superseded by blocker or security bugs discovered post-release, after we've passed that critical marker.

i don't think closing the issue is necessary/helpful, as you highlighted earlier in the issue we need to (and do) consider multiple releases and timelines. no one should feel burdened by these discussions -- but keeping them open makes it clear they are not resolved.

**Comment by ninavizz** [comment]
> I don't mean "dev journal" type posts, but rather very basic and informative post about the current state of the next release (where we are in the process) in normal user language (not in dev language). it is less even about the content than about maintaining a continuous connection via communication with the userbase. maybe starting with a post every quarter or something.

THIS I love. We have to start messaging more users than just developers, and transitioning (within our human and lean-team limits, heh) towards more consumer-product-like expectations management. 

> i don't think closing the issue is necessary/helpful, as you highlighted earlier in the issue we need to (and do) consider multiple releases and timelines. no one should feel burdened by these discussions -- but keeping them open makes it clear they are not resolved.

Ok... but then I also think there is currently a problem with the +1,000 "issues" repo here, which is that nothing is really a priority. One suggestion: for SecureDrop, we have a kanban board that is maintained by our PM and updated every few days. While it reflects a tightly-managed project (that Qubes simply doesn't have the resources to achieve just yet) it does give good visibility into what's being worked on NOW, what needs to be reviewed, and what we hope to have worked on in the very near future: https://github.com/orgs/freedomofpress/projects/1

Perhaps developing something like this to track what `4.1` dependencies are, release tasks, and then planning tasks to prepare for `4.2`+ (which I would consider this among)?

**Comment by andrewdavidwong** [comment]
> > I don't mean "dev journal" type posts, but rather very basic and informative post about the current state of the next release (where we are in the process) in normal user language (not in dev language). it is less even about the content than about maintaining a continuous connection via communication with the userbase. maybe starting with a post every quarter or something.
> 
> 
> 
> THIS I love. We have to start messaging more users than just developers, and transitioning (within our human and lean-team limits, heh) towards more consumer-product-like expectations management. 
> 
> 

I agree this would be good, but it would also require time from Marek. Right now, I can't provide this kind of simple user-terms messaging, because I simply don't have the information in order to be able to share it with everyone. At best, I could provide a synopsis of the current state of Qubes-issues, but that's not what people really want to know. 

> 
> > i don't think closing the issue is necessary/helpful, as you highlighted earlier in the issue we need to (and do) consider multiple releases and timelines. no one should feel burdened by these discussions -- but keeping them open makes it clear they are not resolved.
> 
> 
> 
> Ok... but then I also think there is currently a problem with the +1,000 "issues" repo here, which is that nothing is really a priority.

That's just not true. Every issue has a priority label (`P:`). Most are the **default** priority, but we upgrade some to **major** and some of those to **critical** and some of those to **blocker**. We also downgrade some to **minor**. This is documented here: https://www.qubes-os.org/doc/reporting-bugs/

> One suggestion: for SecureDrop, we have a kanban board that is maintained by our PM and updated every few days. While it reflects a tightly-managed project (that Qubes simply doesn't have the resources to achieve just yet) it does give good visibility into what's being worked on NOW, what needs to be reviewed, and what we hope to have worked on in the very near future: https://github.com/orgs/freedomofpress/projects/1
> 
> 
> Perhaps developing something like this to track what `4.1` dependencies are, release tasks, and then planning tasks to prepare for `4.2`+ (which I would consider this among)? 

Another nice idea, but also another thing Marek would have to spend time doing.

**Comment by QUser534** [comment]
Hardware compatibility is my biggest issue. Xen 4.15 supports ryzen 4000 and current xen does not. Qubes basically cannot run anything released in the last year.

I imagine minor xen updates do not cause much issue. I think it would make sense for point releases to update xen and even dom0. That would make qubes much more hardware compatible.

**Comment by ninavizz** [comment]
> 
> > One suggestion: for SecureDrop, we have a kanban board that is maintained by our PM and updated every few days. While it reflects a tightly-managed project (that Qubes simply doesn't have the resources to achieve just yet) it does give good visibility into what's being worked on NOW, what needs to be reviewed, and what we hope to have worked on in the very near future: https://github.com/orgs/freedomofpress/projects/1
> > 
> > Perhaps developing something like this to track what 4.1 dependencies are, release tasks, and then planning tasks to prepare for 4.2+ (which I would consider this among)?
> 
> Another nice idea, but also another thing Marek would have to spend time doing.

@andrewdavidwong Actually, this is one task that I feel would probably be best tackled as a synchronous workshop by the entire core team. Which I say, knowing there is great reservation among everyone, to commit to hours-long meetings across time zones and verbal language comfort/fluency. Why? Not because Marek lacks capabilities, but because any one individual tackling such a project would likely shape results more informed by personal biases and IC-domain rabbit-holing, than by a reasonable hybrid of team capabilities and user needs. Marta and I have our brains deep in the survey feedback, you and Michael know the community. 

Marek knows the code and architecture better than anyone, and Frederic and Demi and Unman could also be present to discuss feasibility of efforts with Marek, to weigh with priority. 

If it's not a team effort, then a kanban board loses its purpose. Half the objective is to get user needs and community feedback factored in to a project's roadmap, and the other half is everyone's vision—with of course, Marek the team lead.

Marek's gotta be able to function as "a leader" without also being burdened as "the lone dude." If that makes sense. I'd gladly moderate and structure such a workshop, so that it's not just a group free-for-all; and not hours on-end. Breaks, structure, up-front tasks to bring in initially, etc., are all really important.

**Comment by andrewdavidwong** [comment]
> > > One suggestion: for SecureDrop, we have a kanban board that is maintained by our PM and updated every few days. While it reflects a tightly-managed project (that Qubes simply doesn't have the resources to achieve just yet) it does give good visibility into what's being worked on NOW, what needs to be reviewed, and what we hope to have worked on in the very near future: https://github.com/orgs/freedomofpress/projects/1
> > > Perhaps developing something like this to track what 4.1 dependencies are, release tasks, and then planning tasks to prepare for 4.2+ (which I would consider this among)?
> > 
> > 
> > Another nice idea, but also another thing Marek would have to spend time doing.
> 
> @andrewdavidwong Actually, this is one task that I feel would probably be best tackled as a synchronous workshop by the entire core team.

Ah, so not just another thing _Marek_ would have to spend time doing, but something we would _all_ have to spend time doing. :laughing: 

Anyway, I was mainly referring to this part:
> a kanban board that is maintained by our PM and updated every few days

(Which is not to say that it would not be *worth* spending time doing, just that it would, in fact, be yet another thing that he has to spend time doing.)

**Comment by ninavizz** [comment]
@andrewdavidwong Marta insists that Marek is really many people stuffed into one human meatsack. I am very much with you on desiring to whistle-past that hypothesis, while also remaining protective of how the rest of us spend our time, as well. :) 

![giphy](https://user-images.githubusercontent.com/8262612/115124715-83bc9c00-9f78-11eb-9ceb-ff8d314c82f2.gif)

**Comment by ctrlaltf24** [comment]
> @andrewdavidwong Actually, this is one task that I feel would probably be best tackled as a synchronous workshop by the entire core team.

While that would be ideal, the first step is something. Here's a possible solution:
- If someone is committed to closing an issue, they assign it. And while it is assigned that person is responsible for fixing the issue.
- When appropriate, the core team adds issues then imminently self-assigns them when that issue relates to future direction of the project. For example if DamiMarie is working on getting the GUI VM to work with multiple monitors, they would create an issue for that and self-assign it to let the community know in which direction progress is being made
- At this point the present state of qubes-issue is the present state of Qubes Development
- @andrewdavidwong collaborates (or simply approves) a monthly (?) Qubes progress report with some highlights of large issues/features that were added, and the rough direction for the next month written by member(s) of the community that follows qubes development progress (such as deeplow or myself)
- Perhaps a separate section for Qubes Community related progress (such as the QEMU backend)

Does that sound like a feasible solution @andrewdavidwong ? It is your time, and your call, we're just here to help.

**Comment by andrewdavidwong** [comment]
> > @andrewdavidwong Actually, this is one task that I feel would probably be best tackled as a synchronous workshop by the entire core team.
> 
> 
> 
> While that would be ideal, the first step is something. Here's a possible solution: [...]
> 

This discussion is getting really off-topic for this issue. You're taking about _informational_ updates (e.g., Qubes News posts), whereas this issue is about _software_ updates (e.g., packages). I'll respond briefly, but let's _please_ stop the off topic discussion here.

> - If someone is committed to closing an issue, they assign it. And while it is assigned that person is responsible for fixing the issue.
> 
> - When appropriate, the core team adds issues then imminently self-assigns them when that issue relates to future direction of the project. For example if DamiMarie is working on getting the GUI VM to work with multiple monitors, they would create an issue for that and self-assign it to let the community know in which direction progress is being made
> 
> - At this point the present state of qubes-issue is the present state of Qubes Development
> 

This is already how it works. Devs don't self-assign as consistently as you describe, but I fear that constantly riding them to do so could be counterproductive. Discipline is important, but ultimately the process exists to serve the people, not the other way around.

> - @andrewdavidwong collaborates (or simply approves) a monthly (?) Qubes progress report with some highlights of large issues/features that were added, and the rough direction for the next month written by member(s) of the community that follows qubes development progress (such as deeplow or myself)
> 

See my comments above for why I don't think this would work out very well. It either takes up dev time or ends up being something no one really wanted.

> - Perhaps a separate section for Qubes Community related progress (such as the QEMU backend)
> 

This would be the same, but maybe community devs have more time to spare? (Some might, but many don't.) Regardless, I think our current practice of having them provide updates on their respective issues (and occasionally requesting such updates when none have been volunteered for a while) works pretty well.

**Comment by andrewdavidwong** [comment]
Updated issue title to be more actionable.
